File: frontend/src/types/payment.ts
import type { PaymentState, RentState, LateFeeState } from "./states";

export type PaymentCategory = "rent" | "deposit" | "late_fee" | "maintenance" | "utility" | "other";
export type PaymentMethod = "bank_transfer" | "card" | "cash" | "direct_debit" | "cheque";

export interface Payment {
  id: string;
  state: PaymentState;
  leaseId: string;
  rentScheduleId?: string | null;
  organisationId?: string;
  category: PaymentCategory;
  method?: PaymentMethod;
  // Amounts
  amount: number;
  currency: string;
  paidAt?: string;
  // References
  reference?: string | null;
  idempotencyKey?: string | null;
  // Meta
  notes?: string;
  createdAt: string;
  updatedAt: string;
  // Adaptive payment fields (v4 skill)
  failureReason?: string | null;
  retryCount?: number;
  predictedFailureScore?: number | null;
  recommendedChannel?: string | null;
  // Rejection audit (set by org staff)
  rejectionReason?: string | null;
  rejectedAt?: string | null;
  rejectedByProfileId?: string | null;
  // Cancellation audit (set by tenant)
  cancellationReason?: string | null;
  cancelledAt?: string | null;
  // Denormalised display names
  tenantName?: string | null;
  unitName?: string | null;
  propertyName?: string | null;
}

// ── Adaptive payment types (v4 skill) ─────────────────────────────────────────

export interface ChannelCostEstimate {
  channel: string;
  feePercent: number;
  feeAmount: number;
  totalAmount: number;
}

export interface PaymentDecision {
  recommendedChannel: string;
  predictedFailureScore: number;
  retryStrategy: "none" | "immediate" | "delayed" | "next_day";
  costEstimates: ChannelCostEstimate[];
  explain: string;
}

export interface PaymentEstimateRequest {
  amount: number;
  currency?: string;
  tenantId?: string;
}

export interface RentSchedule {
  id: string;
  reference?: string | null;
  organisationId: string;
  leaseId: string;
  // status is the canonical field from the backend (pending | overdue | paid | waived)
  status: string;
  // state is kept as a read-only alias populated by mapRentSchedule for legacy compat
  state?: string;
  // Period
  periodStart: string;
  periodEnd: string;
  dueDate: string;
  amountDue: number;
  amountPaid: number;
  lateFeeApplied: number;
  balance: number;
  paidAt?: string | null;
  notes?: string | null;
  createdAt: string;
  updatedAt: string;
  // Enriched display names (populated on org-level list queries)
  tenantName?: string | null;
  unitName?: string | null;
  propertyName?: string | null;
}

export interface LateFee {
  id: string;
  state: LateFeeState;
  rentScheduleId: string;
  tenantId: string;
  leaseId: string;
  // Amounts
  baseAmount: number;
  feeType: "flat" | "percentage";
  feeValue: number;
  calculatedAmount: number;
  currency: string;
  // Dates
  triggeredAt: string;
  appliedAt?: string;
  paidAt?: string;
  waivedAt?: string;
  waivedBy?: string;
  waiveReason?: string;
  // Linked
  paymentId?: string;
}

export interface Deposit {
  id: string;
  leaseId: string;
  tenantId: string;
  amount: number;
  currency: string;
  receivedAt?: string;
  heldAt?: string; // Deposit protection scheme reference date
  protectionScheme?: string;
  protectionReference?: string;
  returnedAt?: string;
  returnedAmount?: number;
  deductions: DepositDeduction[];
  status: "pending" | "received" | "protected" | "partially_returned" | "returned" | "disputed";
}

export interface DepositDeduction {
  id: string;
  description: string;
  amount: number;
  evidenceUrls: string[];
  approvedAt?: string;
}

export interface LedgerEntry {
  id: string;
  date: string;
  description: string;
  category: PaymentCategory;
  debit: number;
  credit: number;
  balance: number;
  currency: string;
  paymentId?: string;
  reference?: string;
}

// New allocation-layer types

export interface LedgerEntryV2 {
  id: string;
  organisationId: string;
  leaseId: string;
  entryType: "credit" | "debit";
  amount: number;
  referenceType: string;
  referenceId: string;
  balanceAfter: number;
  description: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface LedgerPage {
  data: LedgerEntryV2[];
  total: number;
  page: number;
  pageSize: number;
  hasNext: boolean;
  currentBalance: number;
}

export interface PaymentAllocation {
  id: string;
  paymentId: string;
  rentScheduleId: string;
  amountApplied: number;
  createdAt: string;
  updatedAt: string;
}

export interface TenantWallet {
  id: string;
  tenantId: string;
  organisationId: string;
  balance: number;
  currency: string;
  createdAt: string;
  updatedAt: string;
}

export interface WalletTransaction {
  id: string;
  tenantId: string;
  organisationId: string;
  transactionType: "credit" | "debit";
  amount: number;
  referenceType: string;
  referenceId: string | null;
  description: string | null;
  balanceAfter: number;
  createdAt: string;
  updatedAt: string;
}

export interface WalletTransactionPage {
  data: WalletTransaction[];
  total: number;
  page: number;
  pageSize: number;
  hasNext: boolean;
}

export type MobileMoneyStatus =
  | "pending"
  | "received"
  | "matched"
  | "unmatched"
  | "failed"
  | "expired";

export interface MobileMoneyTransaction {
  id: string;
  organisationId: string;
  provider: "MTN" | "AIRTEL";
  externalId: string;
  phoneNumber: string;
  amount: number;
  currency: string;
  status: MobileMoneyStatus;
  receivedAt: string | null;
  matchedPaymentId: string | null;
  referenceId: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface MobileMoneyPage {
  data: MobileMoneyTransaction[];
  total: number;
  page: number;
  pageSize: number;
  hasNext: boolean;
}

File: backend/app/schemas/payment.py
"""Pydantic schemas for the payments domain."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import Field, field_validator

from app.schemas.common import CamelModel, PaginatedResponse


# ── Rent Schedule ──────────────────────────────────────────────────────────────

class RentScheduleOut(CamelModel):
    id: str
    reference: str | None = None
    organisation_id: str
    lease_id: str
    period_start: str       # ISO date
    period_end: str         # ISO date
    due_date: str           # ISO date
    amount_due: float
    amount_paid: float
    late_fee_applied: float
    balance: float          # amount_due + late_fee_applied - amount_paid
    status: str
    paid_at: str | None
    notes: str | None
    created_at: str
    updated_at: str
    # Denormalised display names (populated on org-level list queries)
    tenant_name: str | None = None
    unit_name: str | None = None
    property_name: str | None = None


# ── Payment ────────────────────────────────────────────────────────────────────

class PaymentCreate(CamelModel):
    # Optional: if supplied the payment is linked to a specific schedule.
    # If omitted, confirm_payment() allocates automatically (oldest-first).
    rent_schedule_id: str | None = None
    amount: float = Field(gt=0)
    currency: str = "UGX"
    category: str = "rent"
    method: str = "cash"
    phone: str | None = None            # mobile money: triggers STK push
    reference: str | None = None
    receipt_url: str | None = None      # bank transfer / cash proof of payment
    idempotency_key: str | None = None
    paid_at: datetime | None = None     # defaults to now() in service
    notes: str | None = None

    @field_validator("category")
    @classmethod
    def valid_category(cls, v: str) -> str:
        allowed = {"rent", "deposit", "late_fee", "other"}
        if v not in allowed:
            raise ValueError(f"category must be one of {allowed}")
        return v

    @field_validator("method")
    @classmethod
    def valid_method(cls, v: str) -> str:
        allowed = {"cash", "bank_transfer", "mobile_money_mtn", "mobile_money_airtel", "other"}
        if v not in allowed:
            raise ValueError(f"method must be one of {allowed}")
        return v


class PaymentCreateFlat(CamelModel):
    """Used by flat POST /payments — lease_id is supplied in the body."""
    lease_id: str
    rent_schedule_id: str | None = None
    amount: float = Field(gt=0)
    currency: str = "UGX"
    category: str = "rent"
    method: str = "cash"
    phone: str | None = None            # mobile money: triggers STK push
    reference: str | None = None
    receipt_url: str | None = None      # bank transfer / cash proof of payment
    idempotency_key: str | None = None
    paid_at: datetime | None = None
    notes: str | None = None

    @field_validator("category")
    @classmethod
    def valid_category(cls, v: str) -> str:
        allowed = {"rent", "deposit", "late_fee", "other"}
        if v not in allowed:
            raise ValueError(f"category must be one of {allowed}")
        return v

    @field_validator("method")
    @classmethod
    def valid_method(cls, v: str) -> str:
        allowed = {"cash", "bank_transfer", "mobile_money_mtn", "mobile_money_airtel", "other"}
        if v not in allowed:
            raise ValueError(f"method must be one of {allowed}")
        return v


class ManualPaymentCreate(CamelModel):
    """Used by POST /leases/{id}/payments/record to record an out-of-band payment."""
    amount: float = Field(gt=0)
    currency: str = "UGX"
    category: str = "rent"
    method: str = "bank_transfer"
    paid_at: datetime | None = None   # when the tenant actually paid; defaults to now()
    reference: str | None = None      # mobile money / bank transaction reference
    notes: str | None = None

    @field_validator("category")
    @classmethod
    def valid_category(cls, v: str) -> str:
        allowed = {"rent", "deposit", "late_fee", "other"}
        if v not in allowed:
            raise ValueError(f"category must be one of {allowed}")
        return v

    @field_validator("method")
    @classmethod
    def valid_method(cls, v: str) -> str:
        allowed = {"cash", "bank_transfer", "mobile_money_mtn", "mobile_money_airtel", "cheque", "other"}
        if v not in allowed:
            raise ValueError(f"method must be one of {allowed}")
        return v


class PaymentOut(CamelModel):
    id: str
    organisation_id: str
    lease_id: str
    rent_schedule_id: str | None
    amount: float
    currency: str
    category: str
    method: str
    reference: str | None
    idempotency_key: str | None
    status: str
    paid_at: str | None
    notes: str | None
    # Adaptive fields
    failure_reason: str | None = None
    retry_count: int = 0
    predicted_failure_score: float | None = None
    recommended_channel: str | None = None
    # Rejection audit (set when org staff rejects)
    rejection_reason: str | None = None
    rejected_at: str | None = None
    rejected_by_profile_id: str | None = None
    # Cancellation audit (set when tenant cancels)
    cancellation_reason: str | None = None
    cancelled_at: str | None = None
    created_at: str
    updated_at: str
    # Denormalised display names
    tenant_name: str | None = None
    unit_name: str | None = None
    property_name: str | None = None
    # Informational message returned to the client (e.g. "Check your phone for PIN")
    message: str | None = None


class PaymentRejectBody(CamelModel):
    reason: str = Field(min_length=1, description="Why this payment is being rejected")


class PaymentCancelBody(CamelModel):
    reason: str | None = Field(None, description="Optional reason for cancellation (shown to manager)")


# ── Late Fee ───────────────────────────────────────────────────────────────────

class LateFeeOut(CamelModel):
    id: str
    organisation_id: str
    lease_id: str
    rent_schedule_id: str
    fee_type: str
    calculated_amount: float
    applied_at: str
    waived: bool
    waived_at: str | None
    waived_reason: str | None
    created_at: str
    updated_at: str


class LateFeeWaive(CamelModel):
    reason: str = Field(min_length=1)


# ── Deposit ────────────────────────────────────────────────────────────────────

class DeductionItem(CamelModel):
    reason: str = Field(min_length=1)
    amount: float = Field(gt=0)


class DepositReturn(CamelModel):
    amount_returned: float = Field(gt=0)
    deductions: list[DeductionItem] = []
    notes: str | None = None


class DepositOut(CamelModel):
    id: str
    organisation_id: str
    lease_id: str
    amount_held: float
    amount_returned: float
    deductions: list[dict[str, Any]]
    status: str
    returned_at: str | None
    notes: str | None
    created_at: str
    updated_at: str


# ── Ledger ─────────────────────────────────────────────────────────────────────

class LedgerOut(CamelModel):
    lease_id: str
    currency: str
    # Rent
    total_rent_due: float
    total_rent_paid: float
    total_rent_outstanding: float
    overdue_schedules: int
    # Late fees
    total_late_fees: float
    total_late_fees_waived: float
    # Deposit
    deposit_held: float | None
    deposit_returned: float | None
    deposit_status: str | None
    # Payments summary
    total_payments: int
    total_confirmed: float


# ── Payment Allocation ─────────────────────────────────────────────────────────

class PaymentAllocationOut(CamelModel):
    id: str
    payment_id: str
    rent_schedule_id: str
    amount_applied: float
    created_at: str
    updated_at: str


# ── Ledger ─────────────────────────────────────────────────────────────────────

class LedgerEntryOut(CamelModel):
    id: str
    organisation_id: str
    lease_id: str
    entry_type: str           # "credit" | "debit"
    amount: float
    reference_type: str       # "payment" | "refund" | "late_fee" | "deposit" | "overpayment" | "wallet"
    reference_id: str
    balance_after: float
    description: str | None
    created_at: str
    updated_at: str


class LedgerPageOut(CamelModel):
    data: list[LedgerEntryOut]
    total: int
    page: int
    page_size: int
    has_next: bool
    current_balance: float


# ── Wallet ─────────────────────────────────────────────────────────────────────

class WalletOut(CamelModel):
    id: str
    tenant_id: str
    organisation_id: str
    balance: float
    currency: str
    created_at: str
    updated_at: str


class WalletTransactionOut(CamelModel):
    id: str
    tenant_id: str
    organisation_id: str
    transaction_type: str     # "credit" | "debit"
    amount: float
    reference_type: str
    reference_id: str | None
    description: str | None
    balance_after: float
    created_at: str
    updated_at: str


WalletTransactionPageOut = PaginatedResponse[WalletTransactionOut]


# ── Adaptive Payment (v4 skill) ────────────────────────────────────────────────

class ChannelCostEstimateOut(CamelModel):
    channel: str
    fee_percent: float
    fee_amount: float
    total_amount: float


class PaymentEstimateRequest(CamelModel):
    amount: float = Field(gt=0)
    currency: str = "UGX"
    tenant_id: str | None = None   # optional — enables per-tenant failure prediction


class PaymentDecisionOut(CamelModel):
    recommended_channel: str
    predicted_failure_score: float
    retry_strategy: str            # none | immediate | delayed | next_day
    cost_estimates: list[ChannelCostEstimateOut]
    explain: str


# ── Mobile Money ───────────────────────────────────────────────────────────────

class MobileMoneyTransactionOut(CamelModel):
    id: str
    organisation_id: str
    provider: str             # "MTN" | "AIRTEL"
    external_id: str
    phone_number: str
    amount: float
    currency: str
    status: str               # pending | received | matched | unmatched | failed | expired
    received_at: str | None
    matched_payment_id: str | None
    reference_id: str | None
    created_at: str
    updated_at: str


MobileMoneyPageOut = PaginatedResponse[MobileMoneyTransactionOut]


# ── Bulk confirm ───────────────────────────────────────────────────────────────

class BulkConfirmRequest(CamelModel):
    payment_ids: list[uuid.UUID] = Field(
        ...,
        min_length=1,
        max_length=50,
        description="IDs of payments to confirm. Max 50 per request.",
    )


class BulkConfirmFailure(CamelModel):
    id: uuid.UUID
    reason: str


class BulkConfirmResult(CamelModel):
    confirmed: list[PaymentOut]
    failed: list[BulkConfirmFailure]


# ── Wallet credit ──────────────────────────────────────────────────────────────

class WalletCreditRequest(CamelModel):
    amount: float = Field(..., gt=0, description="Amount to credit (positive, in org currency)")
    description: str | None = Field(None, max_length=255)
